### ISSUE TITLE
C API for Subclassing

### ISSUE THREAD

**Issue by Shinmera** [Shirakumo/soloud#3]
A lot of classes should be subclassable through a C API. Specifically, the following superclasses need this functionality extension:

* `SoLoud::AudioSource`
* `SoLoud::Filter`
* `SoLoud::AudioCollider`
* `SoLoud::AudioAttenuator`

This should probably happen by a specific subclass (`VirtualAudioSource`, `VirtualFilter`, etc) to each that in turn allows creating "virtual subclasses" identified by some kind of integer ID. A class instance would take such an ID upon creation and realise the required methods through delegates to the appropriate callbacks.

**Comment by Shinmera** [comment]
Proposed implementation:

VirtualFilter.hpp
```c++
class VirtualFilter : public Filter {
public:
  unsigned int classID;
  static void (*mFilter)(unsigned int, unsigned int, unsigned int, float, time);
  static void (*mFilterChannel)(unsigned int, unsigned int, float, time, unsigned int, unsigned int);

  virtual FilterInstance *createInstance();
  VirtualFilter(unsigned int classID);
};

class VirtualFilterInstance : public FilterInstance {
private:
  VirtualFilter *mParent;

public:
  virtual void filter(float *aBuffer, unsigned int aSamples, unsigned int aChannels,
                      float aSamplerate, time aTime);
  virtual void filterChannel(float *aBuffer, unsigned int aSamples, float aSamplerate,
                             time aTime, unsigned int aChannel, unsigned int aChannels);

  VirtualFilterInstance(VirtualFilter *parent);
};
```
VirtualFilter.cpp
```c++
VirtualFilter::VirtualFilter(unsigned int classID){
  this->classID = classID;
}

FilterInstance *VirtualFilter::createInstance(){
  return new VirtualFilterInstance(this);
}

void VirtualFilterInstance::filter(float *aBuffer, unsigned int aSamples, unsigned int aChannels,
                              float aSamplerate, time aTime){
  this->mParent->mFilter(this->mParent->classID, aBuffer, aSamples, aChannels, aSamplerate, aTime);
}

void VirtualFilterInstance::filterChannel(float *aBuffer, unsigned int aSamples, float aSamplerate,
                                          time aTime, unsigned int aChannel, unsigned int aChannels){
  this->mParent->mFilterChannel(this->mParent->classID, aBuffer, aSamples, aSamplerate, aTime, aChannel, aChannels);
}
```
soloud_c.c
```c++
void *VirtualFilter_create(unsigned int classID){
  return (void *)new VirtualFilter(classID);
}

void VirtualFilter_destroy(void *virtualFilter){
  delete (VirtualFilter *)virtualFilter;
}

void VirtualFilter_setFilterCallback(void *virtualFilter, void (*filter)){
  ((VirtualFilter *)virtualFilter)->mFilter = filter;
}

void VirtualFilter_setFilterChannelCallback(void *virtualFilter, void (*filterChannel)){
  ((VirtualFilter *)virtualFilter)->mFilterChannel = filterChannel;
}
```

**Comment by Shinmera** [comment]
Implemented the above, see b48b739d5b4688a48250266dd2753b204ca1959c

Untested, though.

**Comment by Shinmera** [comment]
Alright, even though we have audio sources subclassable now, there's some other stuff missing, most importantly of which is figuring out how the hell the audio source is supposed to choose its sample rate, and what the fuck those seek arguments are supposed to be.
